lib/pg.utils.js

```diff
@@ -35,8 +35,61 @@ const PREDEFINED_DOMAINS = {
 
 const IGNORED_DOMAINS = ['List', 'Enum', 'HashMap', 'HashSet'];
 
-const isValidIdentifier =
-  name => !'0123456789'.includes(name[0]); // TODO: implement validation
+// https://tools.ietf.org/html/rfc3629#section-3
+const utf8bytesLastCodePoints = {
+  1: 0x007F,
+  2: 0x07FF,
+  3: 0xFFFF,
+};
+
+const utf8codePointSize = codePoint => {
+  for (let byteCount = 1; byteCount <= 3; byteCount++) {
+    if (codePoint <= utf8bytesLastCodePoints[byteCount]) {
+      return byteCount;
+    }
+  }
+  return 4;
+};
+
+const asciiCP = {
+  aToZLower: ['a'.codePointAt(0), 'z'.codePointAt(0)],
+  aToZUpper: ['A'.codePointAt(0), 'Z'.codePointAt(0)],
+  underscore: '_'.codePointAt(0),
+  numbers: ['0'.codePointAt(0), '9'.codePointAt(0)],
+  dollar: '$'.codePointAt(0),
+};
+
+// https://github.com/postgres/postgres/blob/5f6b0e6d69f1087847c8456b3f69761c950d52c6/src/backend/utils/adt/misc.c#L723
+const isValidIdentifierStart = cp => cp === asciiCP.underscore ||
+  (cp >= asciiCP.aToZLower[0] && cp <= asciiCP.aToZLower[1]) ||
+  (cp >= asciiCP.aToZUpper[0] && cp <= asciiCP.aToZUpper[1]) ||
+  cp >= 0x80;
+
+// https://github.com/postgres/postgres/blob/5f6b0e6d69f1087847c8456b3f69761c950d52c6/src/backend/utils/adt/misc.c#L741
+const isValidIdentifierCont =
+  cp => (cp >= asciiCP.numbers[0] && cp <= asciiCP.numbers[1]) ||
+    cp === asciiCP.dollar ||
+    isValidIdentifierStart(cp);
+
+const singleUnitUtf16 = 1 << 16 - 1;
+
+// https://www.postgresql.org/docs/10/static/sql-syntax-lexical.html#SQL-SYNTAX-IDENTIFIERS
+const isValidIdentifier = name => {
+  let length = 0;
+  for (let i = 0; i < name.length; i++) {
+    const codePoint = name.codePointAt(i);
+    if (i === 0) {
+      if (!isValidIdentifierStart(codePoint)) {
+        return false;
+      }
+    } else if (!isValidIdentifierCont(codePoint)) {
+      return false;
+    }
+    length += utf8codePointSize(codePoint);
+    if (codePoint > singleUnitUtf16) i++;
+  }
+  return length < 64;
```

Comment:
This limit can be raised/changed by changing PG settings, maybe it's worth making this an argument to this function with the default being 64?

Reply:
Changing this limit requires full PostgreSQL recompilation (which I doubt will be done at any point). Besides, this function is supposed to be used to validate schemas before initializing the corresponding database structure, and I don't think it will be possible to determine the value of this constraint at that point.

Reply:
Why would we need to determine it? If PG was recompiled with a custom value then call to this function will have to specify that value, it's just a matter of where the constant is set.

Reply:
Because if we can't determine it, there is no way to correctly validate the identifiers, since the value provided by the user can differ from the value PostgreSQL was compiled with.

Reply:
Well, as I mentioned it's only a matter of where the constant is specified: here, or by a callee. (If PG was recompiled with a different constant then a callee will know that and put that new constant in a call to this function)

Reply:
@lundibundi currently this argument will not be used. Also I don't think we'll such identifiers. I'd like to land this ASAP, so if you insist this can be done in a follow-up PR.

Reply:
I don't this was just a suggestion/topic to discuss.